Modify S1121: Migrate to LayC - Assignments should not be made from within sub-expressions (#3269)
SonarSource/rspec: rules/S1121/cfamily/rule.adoc

== Why is this an issue?

Assigning a value inside a condition (of an `if` statement, a `for` statement, a `while` or a `switch`) can be confusing. It assigns the value and checks it at the same time, but it is easily confused with a simple equality check with `==` and the original intention can be unclear.

[source,cpp]
----
void action(int parm) {
  if (parm = getValue()) { // Noncompliant: assigning and checking. Is it on purpose?
    doSomething();
  }
  // ...
}
----

{cpp}17 introduced `if` and `switch` statements with initializers, which allows to clearly separate the assignment and the check while keeping a compact code:

[source,cpp]
----
void action(int parm) {
  if (parm = getValue(); parm) { // Compliant: Assignment and check are separate
    doSomething();
  }
  // ...
}
----

> For older versions of the standard, it's better to assign before the statement:

Reviewer: We should avoid contractions.
```suggestion
For older versions of the standard, it is better to assign before the statement:
```

> For older versions of the standard, it's better to assign before the statement:

Reviewer: @frederic-tingaud-sonarsource Isn't this version the best even in C++17+? My thinking is that the purpose of the init statement in if/switch is to introduce a variable with a limited scope, which is different from the proposed fix. WDYT?

So, I would start with the C++98 solution. If you think is has value, I would mention it's possible to use C++17 init statement in if/switch without giving an example, but I'm not 100% sure it's interesting to know that in this context.

[source,cpp]
----
void action(int parm) {
  parm = getValue();
  if (parm) { // Compliant
    doSomething();
  }
  // ...
}
----

=== Exceptions

> This rule ignores assignments explicitly enclosed in parentheses.

Reviewer: @frederic-tingaud-sonarsource Should we mention S1110 "Redundant pairs of parentheses should be removed" here? Or do you think it's a detail and would just add noise?

(If relevant, this suggestion is probably not the best...)
```suggestion
This rule ignores assignments explicitly enclosed in parentheses and is compatible with S1110.
```

include::../description.adoc[]

=== Exceptions

This rule ignores assignments explicitly enclosed in parentheses.

[source,cpp]
----
while ((run = keepRunning())) {
  //...
}
----

== How to fix it

include::../how-to-fix-it.adoc[]

=== Code examples

==== Noncompliant code example

[source,cpp,diff-id=1,diff-type=noncompliant]
----
int arr[] = {0, 1, 2, 3, 4};

> if (int a; a = arr[1]) {

Author: Apparently, the logic for CFamily is quite different from the other languages, and the generic example that used to reside in `[non]compliant.adoc` is in fact allowed by sonar-cpp. I tried to come up with a reasonable yet simple example, but it may be not the best one.

== Resources

=== Documentation

* CWE - https://cwe.mitre.org/data/definitions/481[481: Assigning instead of Comparing]

* {cpp} Reference - https://en.cppreference.com/w/cpp/language/if#If_statements_with_initializer[If statements with initializer]

=== Standards

> * CERT https://wiki.sei.cmu.edu/confluence/x/ZNYxBQ[EXP45-C.: Do not perform assignments in selection statements]

Reviewer: ```suggestion
* CERT - https://wiki.sei.cmu.edu/confluence/x/ZNYxBQ[EXP45-C. Do not perform assignments in selection statements]
```

> * CERT https://wiki.sei.cmu.edu/confluence/x/ITZGBQ[EXP51-J.: Do not perform assignments in conditional expressions]

Reviewer: This is the Java version. It's not relevant in this language.
```suggestion
```

> === External guidelines

Reviewer: ```suggestion
=== External coding guidelines
```
